adapter/statsd/statsd.go

```diff
@@ -0,0 +1,161 @@
+	if err != nil {
+		// TODO: something other than panic?
+		panic("failed to construct statsd client with err: " + err.Error())
+	}
+
+	templates := make(map[string]*template.Template)
+	for metricName, s := range conf.MetricNameTemplateStrings {
+		if t, err := template.New(metricName).Parse(s); err != nil {
+			// This should never happen given ValidateConfig, but we call this before validateconfig? unclear how this needs to work.
+			// We could migrate this parsing into a sync.Once and called during NewMetrics, but then it happen during a request.
+			panic(fmt.Errorf("failed to parse template '%s' for metric '%s' with err: %s", s, metricName, err))
+		} else {
+			templates[metricName] = t
+		}
+	}
+
+	return &builder{adapter.NewDefaultBuilder(name, desc, conf), client, templates}
+}
+
+func (b *builder) ValidateConfig(c adapter.AspectConfig) (ce *adapter.ConfigErrors) {
+	params, ok := c.(*config.Params)
+	if !ok {
+		// TODO: should this just report an error?
+		panic(fmt.Errorf("invalid config type, want adapter/statsd/config/config.proto, got: %v", params))
+	}
+
+	// TODO: we need to validate that the provided template strings are satisfiable by the mixer's attributes/metric's labels
+	for metricName, s := range params.MetricNameTemplateStrings {
+		if _, err := template.New(metricName).Parse(s); err != nil {
+			ce = ce.Append("template "+metricName, fmt.Errorf("failed to parse template '%s' for metric '%s' with err: %s", s, metricName, err))
+		}
+	}
+	return
+}
+
+func (b *builder) Close() error {
+	return b.client.Close()
+}
+
+func (b *builder) NewMetricsAspect(env adapter.Env, cfg adapter.AspectConfig, metrics []adapter.MetricDefinition) (adapter.MetricsAspect, error) {
+	// TODO: should we verify that templates are satisfiable here? we have the labels from the definitions, but it's not clear other than
+	// trying to fill in each template that we can satisfy them.
+
+	params, ok := cfg.(*config.Params)
+	if !ok {
+		return nil, fmt.Errorf("invalid config, expected statsd.config.Params, got %+v", cfg)
+	}
+	return &aspect{params.SamplingRate, b.client, b.templates}, nil
+}
+
+func (a *aspect) Record(values []adapter.Value) error {
+	for _, v := range values {
+		if err := a.record(v); err != nil {
+			// TODO: should we keep track of all of the errors and return a composite, rather than aborting at the first problem?
+			return err
+		}
+	}
+	return nil
+}
+
+func (a *aspect) record(value adapter.Value) error {
+	name := value.Name
+	if t, found := a.templates[value.Name]; found {
+		buf := new(bytes.Buffer)
+		if err := t.Execute(buf, value.Labels); err != nil {
+			// TODO: we should be able to verify that this will never happen at config time, should this panic?
+			return fmt.Errorf("failed to create metric name with template '%s' and labels '%v'; got err: %s", t.Name(), value.Labels, err)
+		}
+		name = buf.String()
+	}
+
+	switch value.Kind {
+	case adapter.Gauge:
+		v, err := value.Int64()
```

Comment:
And based on my reading, statsd distributions (histograms) are configured on the collection side, not the publishing side. So the user can configure their server to retain metric "foo" as a histogram, and configure Istio to report a time for metric "foo", which their collection server will retain as a histogram.

As far as supporting timing and other metric types, we just need to add more metric kinds; I didn't want to conflate adding more kinds with the PR to add the adapter.